[JSC] JSPI Implementation

ast-hugger opened the pull request:
#### 53e97afd3421bd259df1beeb7d58f7d33d9c8b63
<pre>
[JSC] JSPI Implementation

Reviewed by Yusuke Suzuki.

This implements the WebAssembly JavaScript Promise Integration proposal:
<a href="https://github.com/WebAssembly/js-promise-integration">https://github.com/WebAssembly/js-promise-integration</a>

Controlled by the feature flag `useJSPI`, disabled by default.

The proposal adds a new API function and a constructor to the WebAssembly object:

  - WebAssembly.promising(wasmFun)
  - new WebAssembly.Suspending(jsFun)

Each creates a wrapper function with special behavior around the function passed as the argument.
We will refer to these as a &quot;promising wrapper&quot; and a &quot;suspending wrapper&quot;.

The behavior of a promising wrapper is implemented by runWebAssemblyPromisingFunction() in
WebAssemblyPromising.cpp. The wrapper does the following:

  - Creates a promise which will later be resolved or rejected based on the execution outcome
    of the wrapped function (&quot;result promise&quot;).

  - Makes arrangements so that if a suspending wrapper is invoked during the call to the wrapped
    function, the suspending wrapper can locate the stack frame associated with the promising
    wrapper.

  - Calls the wrapped function (&apos;wasmFun&apos;).

The behavior of a suspending wrapper is implemented by runWebAssemblySuspendingFunction() in
WebAssemblySuspending.cpp, though the actual entry point is an offlineasm function
&apos;enterWebAssemblySuspendingFunction&apos; in InPlaceInterpreter.asm. The separate entry function is
necessary to capture callee-save registers on entry, before the C++ function prologue can tamper
with them.

The suspending wrapper:

  - Calls the wrapped function (&apos;jsFun&apos;). If the value returned by the call is not a promise,
    creates a promise (&quot;suspending promise&quot;) and resolves it with the returned value.
    Otherwise, the promise returned by jsFun is the suspending promise.

  - Captures the stack between its own frame and the frame of the nearest promising wrapper
    invocation as a heap object (class EvacuatedStack).

  - Creates an object (class PinballCompletion) that bundles together the evacuated stack
    representing the suspended Wasm computation and the result promise.

  - Sets a fulfillment handler on the suspending promise. When the promise is
    resolved, the handler will take the resolution result and feed it into the suspended
    Wasm computation. To do that, the handler will &quot;implant&quot; some or all of the previously
    evacuated stack frames onto the stack of the microtask it is executing on, and return
    into those frames. (&quot;Pinball&quot; in the name comes from the ability to not implant all
    of the evacuated frames at once, instead returning down the evacuated Wasm stack
    in a series of &quot;bumps&quot;). When the entire evacuated Wasm stack returns with the result,
    that result is used to resolve the result promise.

  - Also sets a rejection handler on the suspending promise. If the rejection reason is NOT
    a Wasm exception, the evacuated Wasm stack is discarded and the rejection reason is used to
    reject the result promise. If the rejection reason is a Wasm exception, the evacuated Wasm stack
    is implanted back onto the current stack and the exception is propagated down that stack.

  - Finally, the suspending wrapper &quot;teleports&quot; into the frame of the promising wrapper by returning
    so that the evacuated stack frames are bypassed.

Fulfillment and rejection handlers of the suspending promise are implemented in offlineasm
and C++ in InPlaceInterpreter.asm and PinballCompletion.cpp.

From the perspective of the promising wrapper, a teleport from the suspending wrapper appears as a
return from the wrapped function. The wrapped function can also return as a result of Wasm
computation completing without ever suspending, or as a result of an exception thrown in Wasm code.
The wrapper checks which scenario is the case. In cases of a normal or exception return, it fulfills
or rejects the result promise. In the case of a teleport, nothing needs to be done because the
result promise is already hooked up to be resolved or rejected by PinballCompletion.

Testing:
  - added multiple stress tests (JSTests/wasm/stress/jspi-*.js)
  - enabled w3c tests under LayoutTests/imported/w3c/web-platform-tests/wasm/jsapi/jspi

Canonical link: <a href="https://commits.webkit.org/307271@main">https://commits.webkit.org/307271@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e08a4647fd990c1a9cbdf8288066bd75e50a4bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/143852 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/16521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/8043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/152520 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/97091 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/1ce617c2-d0c4-4cde-a205-58e81f1cbd29) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/17010 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/16421 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/110617 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/79561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/2ac92e09-2143-4955-b766-5599007df317) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/146815 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/13049 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/129256 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/91535 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/92adf0f7-c21f-4aec-9cb9-f6da9ff6796f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/12514 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/10245 "Passed tests") | [⏳ 🛠 gtk3-libwebrtc ](https://ews-build.webkit.org/#/builders/GTK-GTK3-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/135840 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/121976 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/5879 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/154832 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/4658 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/16381 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/6922 "Found 1 new test failure: fast/images/mac/play-pause-individual-animation-context-menu-items.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/118628 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/16416 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/13771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/118983 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/30503 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/14903 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/127088 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/71794 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/16002 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/5573 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/175138 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/15736 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/79773 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/45175 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/15948 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/15801 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

syg commented on Source/JavaScriptCore/runtime/EvacuatedStack.cpp:
```diff
@@ -0,0 +1,469 @@
+        }
+#else
+        resignedAddr = isTopFrame ? returnPC : frameRecord->returnPC;
+        UNUSED_VARIABLE(isTopFrame);
```
Quick drive-by: `isTopFrame` is actually used now.
```suggestion
```

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyPromisingFunction.h:
```diff
@@ -0,0 +1,68 @@
+
+// A function object returned by a call to 'WebAssembly.promising(f)'.
+class JSWebAssemblyPromisingFunction final : public JSFunction {
+public:
```
In this case, you do not need to create a new function class. You can use JSFunctionWithField so that you do not need a new class (and new IsoSubspace).

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblySuspendingFunction.h:
```diff
@@ -0,0 +1,68 @@
+namespace JSC {
+
+// A function object returned by 'new WebAssembly.Suspending(f)'.
+class JSWebAssemblySuspendingFunction final : public JSFunction {
```
Ditto.

ast-hugger replied:
Done, this and the promising function. (And also the suspending promise handler in PinballCompletion.cpp).

kmiller68 commented on Source/JavaScriptCore/runtime/EvacuatedStack.cpp:
```diff
@@ -0,0 +1,496 @@
+std::unique_ptr<EvacuatedStackSlice> EvacuatedStackSlice::create(std::span<Register> stackSpan, Vector<unsigned>&& frameOffsets, const void* entryPC)
+{
+    ASSERT(stackSpan.size());
+    ASSERT(!(stackSpan.size() % 2)); // In asm code we assume that slice byte sizes are stack-aligned
```
Nit: I think this would be clearer as `ASSERT(isMultipleOf(stackAlignmentRegisters(), stackSpan.size()));` or `ASSERT(roundUpToMultipleOf(stackAlignmentRegisters(), stackSpan.size()) == stackSpan.size());` since `isMultipleOf` doesn't seem to exist in MathExtras.h (although I think it would be a reasonable thing to add).

kmiller68 commented on Source/JavaScriptCore/runtime/EvacuatedStack.cpp:
```diff
@@ -0,0 +1,496 @@
+{
+    return !(reinterpret_cast<uintptr_t>(pointer) & (stackAlignmentBytes() - 1));
+}
+
```
I would add this to StackAlignment.h as it seems like the kind of thing others may want to use in the future (and probably implement it with `roundUpToMultipleOf`/ a new `isMultipleOf`)

kmiller68 commented on Source/JavaScriptCore/runtime/EvacuatedStack.cpp:
```diff
@@ -0,0 +1,496 @@
+    return lastFrame;
+}
+
+// Resigns all return PCs in the saved data after the slice is created. The signing
```
Nit: Re-signs. Resigns sounds like we're giving up :)
```suggestion
// Re-signs all return PCs in the saved data after the slice is created. The signing
```

ast-hugger replied:
LOL, I actually had it named reSignReturnPCs at some point, but then changed everything to match the style of PAC `ptrauth_auth_and_resign`.

kmiller68 commented on Source/JavaScriptCore/runtime/EvacuatedStack.cpp:
```diff
@@ -0,0 +1,496 @@
+// We save this many extra slots above the actual frame record (the fp/lr pair) of a Wasm
+// frame because IPInt stores register values there before a call. Some frame types do not
+// actually use this many slots, but it appears tiering up breaks without a consistent headroom.
+constexpr unsigned StandardHeadroom = 8;
```
```suggestion
constexpr unsigned standardHeadroom = 8;
```

should be camelCase, only types are CamelCase. 

We should also indicate this in some shared location rather than just a comment here. This would be nearly impossible to find if refactoring IPInt in some way so this value should be located somewhere such an author would have to look.

kmiller68 commented on Source/JavaScriptCore/runtime/EvacuatedStack.cpp:
```diff
@@ -0,0 +1,496 @@
+        case Wasm::CompilationMode::IPIntMode:
+        case Wasm::CompilationMode::BBQMode:
+        case Wasm::CompilationMode::OMGMode:
+            return callFrame->registers() + 2;
```
Doesn't this need to include the number of arguments on the stack? At least for OMG they can be read from the frame by the function body so they have to be restored. I think that for IPInt/BBQ/WasmToJS they're copied into the locals though (or are unused for WasmToJS) so this might be correct for those modes.

Can we add some test cases that use 20+ arguments either way? I don't see any in the tests above and it would be good additional coverage for essentially all the cases you added.

ast-hugger replied:
We are now locked into the 'slab' slicing strategy that always saves Wasm stack as a single span, so this is not an issue. I did add a test case that has a call with 24 arguments in the middle of the stack. Surprisingly, it seems to work even with frame-by-frame slicing. We'll shake it out what's going as we polish.

kmiller68 reviewed (commented):
Posting initial comemnts will keep reviewing.

Constellation commented on Source/JavaScriptCore/runtime/EvacuatedStack.h:
```diff
@@ -0,0 +1,237 @@
+{
+#if CPU(ARM64E)
+    return ptrauth_auth_data(m_entryPC, ptrauth_key_asib, this);
+#else
```
Due to JITCage, probably you need to just remove tagged bits. auth cannot work.

ast-hugger replied:
I restructured EvacuatedStack, this piece is now gone. All re-signing is now done in JITCage-friendly way by a single helper function.

Constellation commented on Source/JavaScriptCore/runtime/PinballCompletion.h:
```diff
@@ -0,0 +1,117 @@
+// returning straight down all captured Wasm frame, we may do so in a series of bumps
+// as we execute evacuated slices one after another.
+
+class PinballCompletion : public JSNonFinalObject {
```
Put `final` when it is the final class.

Constellation commented on Source/JavaScriptCore/runtime/PinballCompletion.cpp:
```diff
@@ -0,0 +1,315 @@
+    ASSERT_GC_OBJECT_INHERITS(thisObject, info());
+    Base::visitChildren(thisObject, visitor);
+    if (thisObject->m_resultPromise)
+        visitor.append(thisObject->m_resultPromise);
```
This check is not necessary. You can just do `visitor.append(thisObject->m_resultPromise);`

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
```diff
@@ -377,6 +386,20 @@ JSC_DEFINE_HOST_FUNCTION(webAssemblyInstantiateFunc, (JSGlobalObject* globalObje
+    if (!arg.isCell() || !arg.asCell()->inherits<WebAssemblyFunctionBase>())
+        return JSValue::encode(throwTypeError(globalObject, scope, "Argument 0 must be a WebAssembly exported function"_s));
+
+    JSFunction* wrapped = jsCast<JSFunction*>(arg);
```
You can just do.
```
auto* wrapped = jsDynamicCast<WebAssemblyFunctionBase*>(arg);
if (!wrapped) [[unlikely]]
    return JSValue::encode(throwTypeError(globalObject, scope, "Argument 0 must be a WebAssembly exported function"_s));

RELEASE_AND_RETURN(scope, JSValue::encode(createWebAssemblyPromisingFunction(vm, globalObject, wrapped)));
```

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
```diff
@@ -122,6 +126,11 @@ void JSWebAssembly::finishCreation(VM& vm, JSGlobalObject* globalObject)
         JSC_BUILTIN_FUNCTION_WITHOUT_TRANSITION("instantiateStreaming"_s, webAssemblyInstantiateStreamingCodeGenerator, static_cast<unsigned>(0));
     JSC_NATIVE_GETTER_WITHOUT_TRANSITION("JSTag"_s, webAssemblyGetterJSTag, PropertyAttribute::ReadOnly);
+    if (Options::useJSPI()) {
+        JSC_NATIVE_GETTER_WITHOUT_TRANSITION("Suspending"_s, webAssemblyGetterSuspending, PropertyAttribute::DontEnum);
+        JSC_NATIVE_GETTER_WITHOUT_TRANSITION("SuspendError"_s, webAssemblyGetterSuspendError, PropertyAttribute::DontEnum);
```
Is it actually a getter? Or just a property?

ast-hugger replied:
Ah, yes, I can make it just a property.

ast-hugger replied:
Ok, this is actually not completely straightforward. The SuspendError constructor to put into the property is not available yet at the time this runs. I'll leave it as is for now, if there is a better way we can address it during polishing.

kmiller68 replied:
In what sense? You can allocate other objects inside `finishCreation` generally. You just can't allocate other objects in the constructors.

ast-hugger replied:
Will address this in a follow-up PR.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblySuspending.cpp:
```diff
@@ -0,0 +1,193 @@
+
+namespace JSC {
+
+extern "C" void teleportOverEvacuatedFrames(CPURegister* calleeSaves, CallFrame* returnOutOfFrame, EncodedJSValue result);
```
Let's put SYSV_ABI.

syg commented on Source/JavaScriptCore/runtime/EvacuatedStack.h:
```diff
@@ -0,0 +1,230 @@
+
+#pragma once
+
+#include "interpreter/CallFrame.h"
```
```suggestion
#include <JavaScriptCore/CallFrame.h>
```

EvacuatedStack.h is a Private header, so using `"interpreter/CallFrame.h"` will fail to module verify and build internally.

ast-hugger replied:
I wonder how this slipped in, I never write includes like this. Does clangd do auto-includes in this style? Fixed.

syg replied:
Oh yeah clangd has the annoying header-insertion on by default, be sure to turn it off with `--header-insertion=never`. I have this in my VSCode:

```
    "clangd.arguments": [
        "--header-insertion=never"
    ],
```

syg commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -1272,6 +1302,9 @@ class VM : public ThreadSafeRefCountedWithSuppressingSaferCPPChecking<VM> {
     Lock m_scratchBufferLock;
     Vector<ScratchBuffer*> m_scratchBuffers;
     size_t m_sizeOfLastScratchBuffer { 0 };
+    Lock m_evacuatedStacksLock;
```
Is this lock needed? The 2 vectors below are processed as roots when the world is stopped (i.e. mutator is stopped) IIUC.

ast-hugger replied:
I used m_scratchBufferLock as a model, though the world is stopped reasoning makes sense to me now. Let's get a second opinion to be sure.

kmiller68 replied:
Yeah, I don't think this lock is needed.

ast-hugger replied:
Will address this in a follow-up PR.

syg commented on Source/JavaScriptCore/runtime/Gate.h:
```diff
@@ -50,6 +50,9 @@ namespace JSC {
     v(vmEntryToJavaScript, JSEntryPtrTag) \
+    v(relocateJITReturnPC, NoPtrTag) \
+    v(exitImplantedSliceGate, NoPtrTag) \
+    v(getSentinelFrameReturnPCGate, NoPtrTag) \
```
Please move these gates out of the middle between the LLInt utility gates and the JS opcode gates. Those gates are hot and we should keep them as close together as possible for cache lines.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblySuspendingPrototype.cpp:
```diff
@@ -0,0 +1,58 @@
+{
+    Base::finishCreation(vm);
+    ASSERT(inherits(info()));
+}
```
If your finishCreation does not do anything special, you can just use DECLARE_DEFAULT_FINISH_CREATION. See the existing usage of this.

Constellation commented on Source/JavaScriptCore/runtime/EvacuatedStack.cpp:
```diff
@@ -0,0 +1,496 @@
+#if CPU(ARM64E)
+    auto* originalSignatureSP = reinterpret_cast<const void*>(originalFP + 1);
+    auto* newSignatureSP = reinterpret_cast<const void*>(newFP + 1);
+    return isJITPC(ptrauth_strip(returnPC, 0))
```
Use `removeCodePtrTag`

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblySuspending.cpp:
```diff
@@ -0,0 +1,193 @@
+    if (!promise) {
+        // The spec requires us to suspend even if the wrapped function returned a real value.
+        promise = JSPromise::create(vm, globalObject->promiseStructure());
+        promise->resolve(globalObject, result);
```
This can throw an error. So check is necessary (You can pass thenable object).

kmiller68 replied:
Yeah, we should add a test for this.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblySuspending.cpp:
```diff
@@ -0,0 +1,193 @@
+    // with profile data associated with the promising wrapper. For now the default strategy is
+    // 'slab', and it's the only one that does the right thing for exceptions. The strategy choice
+    // option is for experimentation only.
+    const char* strategy = Options::slicingStrategyForJSPI();
```
String comparison is costly, so at least, we should use integer or something instead.

kmiller68 replied:
Yeah, there are other options that use enums. We should probably do the same here.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblySuspending.cpp:
```diff
@@ -0,0 +1,193 @@
+#if ASAN_ENABLED
+    __asan_handle_no_return(); // the teleport call makes this function not return normally
+#endif
+    teleportOverEvacuatedFrames(vmEntryFrameCalleeSaves, returnOutOfFrame, JSValue::encode(jsUndefined()));
```
This is switching inside C++ function. Isn't it dangerous since C++ things can require some destructor calls etc.?

kmiller68 replied:
Yeah, this should probably return `returnOutOfFrame` and have the wrapper jump to `teleportOverEvacuatedFrames`

Constellation commented on Source/JavaScriptCore/runtime/PinballCompletion.cpp:
```diff
@@ -0,0 +1,316 @@
+    Base::visitChildren(thisObject, visitor);
+
+    visitor.append(thisObject->m_resultPromise);
+    // Evacuated stack slices are registered with the VM and are added to conservative roots,
+    // so no need to do anything about them here.
```
Logically, I think we need to do scanning of this here because we would like to keep evacuated stack scanned only when `PinballCompletion` is marked. Otherwise, if our promise is not resolved and go away, then we will keep evacuated stack forever.
But difficult part is that ConservativeRoots scanning needs to be done non-parallel manner. And conservative scan constraint runs only once per execution stop.
I guess you need to define a new GC constraint, which is not Parallel, and `onstraintVolatility::GreyedByMarking` (executed everytime something gets marked), and do scanning Pinballs's evacuated stack only when Pinball gets marked.
Like, for example, you can have a set in VM (with a lock). This set is cleared when beginMarking. And Pinball's visitor pushes it to this set. And each EvacuatedStack has a bool flag whether it is marked before. And constraint will scan them if it is not having this bool flag (and set a flag). And then, after GC gets done, accessing this set and clear marked flag.

ast-hugger replied:
OK, will address this in a follow up PR.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblySuspending.cpp:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    CPURegister* vmEntryFrameCalleeSaves = vmEntryRecord(vm.topEntryFrame)->calleeSaveRegistersBuffer;
+    memcpySpan(std::span<CPURegister>(vmEntryFrameCalleeSaves, NUMBER_OF_CALLEE_SAVES_REGISTERS), std::span(hereCalleeSaves, NUMBER_OF_CALLEE_SAVES_REGISTERS));
```
Probably worth a FIXME to directly copy to calleeSaveRegistersBuffer in the .asm rather than do it here.

ast-hugger replied:
There is more going on here. In .asm we copy callee saves into a buffer on the stack. That captures callee saves state on suspension. We copy and use that captured state in two places in C++. First place is here, where it's the initial state for unwinding, producing the callee save state in the VM entry frame we use for teleport. The other place is where PinballCompletion is created. We copy the initial callee save state into the pinball object. Later it's used as the callee saves state to restart Wasm execution. So we need that original callee saves state captured on the stack, and both copies.

kmiller68 commented on Source/JavaScriptCore/runtime/PinballCompletion.cpp:
```diff
@@ -0,0 +1,316 @@
+    context->globalObject = globalObject;
+    context->vm = &vm;
+    context->handler = self;
+    new (&context->jspiContext) JSPIContext(JSPIContext::Purpose::Completing, vm, callFrame, pinball->resultPromise());
```
Is there an advantage to placement new here rather than `context->jspiContext = JSPIContext(JSPIContext::Purpose::Completing, vm, callFrame, pinball->resultPromise());`?

ast-hugger replied:
Yes, the constructor of JSPIContext registers itself as the `vm.topJSPIContext`, so it should be constructed in place.

kmiller68 commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -230,6 +232,26 @@ class ActiveScratchBufferScope {
 
+// A record marking the top (in memory address terms) of the "interesting" stack span when
+// handling a JSPI suspension. A pointer to it is saved in the VM as the 'topJSPIContext' field.
+struct JSPIContext {
```
This seems like it should be in its own header as most people looking at VM don't need the details of this struct.

ast-hugger replied:
Ok, will fix it in a follow-up PR.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyPromising.cpp:
```diff
@@ -0,0 +1,102 @@
+JSFunctionWithFields* createWebAssemblyPromisingFunction(VM& vm, JSGlobalObject* globalObject, JSFunction* wrappedFunction)
+{
+    const String name = "<promising wrapper>"_s;
+    NativeExecutable* executable = vm.getHostFunction(runWebAssemblyPromisingFunction, ImplementationVisibility::Public, NoIntrinsic, callHostFunctionAsConstructor, nullptr, name);
```
Should this wrapper function be publicly visible in stack traces? Do you know if it's visible for V8/SM?

ast-hugger replied:
Will make a bugzilla to track this.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblySuspending.cpp:
```diff
@@ -0,0 +1,210 @@
+JSFunctionWithFields* createWebAssemblySuspendingFunction(VM& vm, JSGlobalObject* globalObject, JSValue callable)
+{
+    const String name = "WebAssembly.Suspending"_s;
+    NativeExecutable* executable = vm.getHostFunction(enterWebAssemblySuspendingFunction, ImplementationVisibility::Public, NoIntrinsic, callHostFunctionAsConstructor, nullptr, name);
```
Ditto on visibility.

kmiller68 commented on Source/JavaScriptCore/runtime/PinballCompletion.cpp:
```diff
@@ -0,0 +1,325 @@
+    } else {
+        discardEvacuatedStackAndPropagateRejection(globalObject, pinball, reason);
+        context->exception = nullptr; // this tells the asm driver to exit immediately
+    }
```
We should just throw the exception even if it's a JS exception. That should always "just work" as the unwinder knows which frames can handle JS exceptions or not.

ast-hugger committed:
[JSC] JSPI Implementation

Reviewed by Yusuke Suzuki.

This implements the WebAssembly JavaScript Promise Integration proposal:
https://github.com/WebAssembly/js-promise-integration

Controlled by the feature flag `useJSPI`, disabled by default.

The proposal adds a new API function and a constructor to the WebAssembly object:

  - WebAssembly.promising(wasmFun)
  - new WebAssembly.Suspending(jsFun)

Each creates a wrapper function with special behavior around the function passed as the argument.
We will refer to these as a "promising wrapper" and a "suspending wrapper".

The behavior of a promising wrapper is implemented by runWebAssemblyPromisingFunction() in
WebAssemblyPromising.cpp. The wrapper does the following:

  - Creates a promise which will later be resolved or rejected based on the execution outcome
    of the wrapped function ("result promise").

  - Makes arrangements so that if a suspending wrapper is invoked during the call to the wrapped
    function, the suspending wrapper can locate the stack frame associated with the promising
    wrapper.

  - Calls the wrapped function ('wasmFun').

The behavior of a suspending wrapper is implemented by runWebAssemblySuspendingFunction() in
WebAssemblySuspending.cpp, though the actual entry point is an offlineasm function
'enterWebAssemblySuspendingFunction' in InPlaceInterpreter.asm. The separate entry function is
necessary to capture callee-save registers on entry, before the C++ function prologue can tamper
with them.

The suspending wrapper:

  - Calls the wrapped function ('jsFun'). If the value returned by the call is not a promise,
    creates a promise ("suspending promise") and resolves it with the returned value.
    Otherwise, the promise returned by jsFun is the suspending promise.

  - Captures the stack between its own frame and the frame of the nearest promising wrapper
    invocation as a heap object (class EvacuatedStack).

  - Creates an object (class PinballCompletion) that bundles together the evacuated stack
    representing the suspended Wasm computation and the result promise.

  - Sets a fulfillment handler on the suspending promise. When the promise is
    resolved, the handler will take the resolution result and feed it into the suspended
    Wasm computation. To do that, the handler will "implant" some or all of the previously
    evacuated stack frames onto the stack of the microtask it is executing on, and return
    into those frames. ("Pinball" in the name comes from the ability to not implant all
    of the evacuated frames at once, instead returning down the evacuated Wasm stack
    in a series of "bumps"). When the entire evacuated Wasm stack returns with the result,
    that result is used to resolve the result promise.

  - Also sets a rejection handler on the suspending promise. If the rejection reason is NOT
    a Wasm exception, the evacuated Wasm stack is discarded and the rejection reason is used to
    reject the result promise. If the rejection reason is a Wasm exception, the evacuated Wasm stack
    is implanted back onto the current stack and the exception is propagated down that stack.

  - Finally, the suspending wrapper "teleports" into the frame of the promising wrapper by returning
    so that the evacuated stack frames are bypassed.

Fulfillment and rejection handlers of the suspending promise are implemented in offlineasm
and C++ in InPlaceInterpreter.asm and PinballCompletion.cpp.

From the perspective of the promising wrapper, a teleport from the suspending wrapper appears as a
return from the wrapped function. The wrapped function can also return as a result of Wasm
computation completing without ever suspending, or as a result of an exception thrown in Wasm code.
The wrapper checks which scenario is the case. In cases of a normal or exception return, it fulfills
or rejects the result promise. In the case of a teleport, nothing needs to be done because the
result promise is already hooked up to be resolved or rejected by PinballCompletion.

Testing:
  - added multiple stress tests (JSTests/wasm/stress/jspi-*.js)
  - enabled w3c tests under LayoutTests/imported/w3c/web-platform-tests/wasm/jsapi/jspi

Canonical link: https://commits.webkit.org/307271@main
